File: packages/react-native/android/app/src/main/java/com/netraedge/NetraEdgeModule.kt
package com.netraedge

import android.content.Context
import android.util.Log
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.support.common.FileUtil
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * NetraEdgeModule — SOTA TFLite inference engine for face recognition and liveness.
 *
 * Loads two TFLite models from app assets:
 * - face_recognition.tflite → embedding dim derived at load time
 *   (current pre-trained InsightFace MobileFaceNet = 128-d;
 *    Day-2 fine-tuned EdgeFace-XS / SubCenter-AdaFace / 2-Teacher KD = 512-d).
 * - liveness_detector.tflite → 3-class softmax [real, print, screen]
 *
 * SOTA 2026 stack (Path A+):
 *   - Backbone: EdgeFace-XS gamma-06 (1.77M params, 99.73% LFW paper)
 *   - Loss: SubCenter-AdaFace (K=3) + GRL adversarial debiasing head
 *   - KD: 2 teachers (EdgeFace-base 99.83% LFW + InsightFace 99.38% LFW)
 *   - Embedding: 512-d TARGET (was 128-d MobileFaceNet, 192-d GFN) — code is dim-agnostic
 *   - TTA: horizontal flip averaging at inference time (+0.3-0.5% LFW free)
 *   - Liveness: 7-layer attention fusion (Active + Texture + rPPG + Depth + Depth-Motion + Moiré + Color)
 *
 * Dim handling: the embedding dim is read from the model's output tensor at
 * initialize() time, so the code works with any dim (128, 192, 256, 384, 512, ...).
 * TTA: useTTA defaults to true; set to false to skip the flip forward pass.
 */
class NetraEdgeModule {
    companion object {
        private const val TAG = "NetraEdge"
        private const val INPUT_SIZE = 112
        private const val LIVENESS_CLASSES = 3
        // SOTA TARGET dim for Day-2 fine-tune (Path A+: EdgeFace-XS / 2-Teacher KD).
        // Pre-trained model on disk is 128-d; this is the post-fine-tune target.
        const val SOTA_TARGET_EMBEDDING_DIM = 512
    }

    private var recognitionInterpreter: Interpreter? = null
    private var livenessInterpreter: Interpreter? = null
    private var initialized = false

    /** Embedding dim derived from the loaded recognition model's output tensor. */
    private var embeddingDim: Int = -1

    /** Liveness dim derived from the loaded liveness model's output tensor. */
    private var livenessDim: Int = LIVENESS_CLASSES

    /** Test-time augmentation: horizontal flip averaging at inference. Default ON. */
    private var useTTA: Boolean = true

    fun isInitialized(): Boolean = initialized

    /** Embedding dim of the currently loaded model (or -1 if not loaded). */
    fun getEmbeddingDim(): Int = embeddingDim

    /** Liveness output dim of the currently loaded model. */
    fun getLivenessDim(): Int = livenessDim

    /** Whether TTA (horizontal flip averaging) is enabled. */
    fun isUseTTA(): Boolean = useTTA

    /** Enable or disable TTA at runtime. */
    fun setUseTTA(enabled: Boolean) {
        useTTA = enabled
        Log.i(TAG, "TTA ${if (enabled) "enabled" else "disabled"}")
    }

    fun initialize(context: Context): Boolean {
        return try {
            val recInterpreter = Interpreter(
                FileUtil.loadMappedFile(context, "face_recognition.tflite")
            )
            // Derive output dim from the model's output tensor shape [1, N]
            val recOutShape = recInterpreter.getOutputTensor(0).shape()
            embeddingDim = if (recOutShape.size >= 2) recOutShape[1] else recOutShape.last()
            Log.i(TAG, "Recognition model output dim = $embeddingDim (target = $SOTA_TARGET_EMBEDDING_DIM)")

            val livInterpreter = Interpreter(
                FileUtil.loadMappedFile(context, "liveness_detector.tflite")
            )
            val livOutShape = livInterpreter.getOutputTensor(0).shape()
            livenessDim = if (livOutShape.size >= 2) livOutShape[1] else livOutShape.last()
            Log.i(TAG, "Liveness model output dim = $livenessDim")

            recognitionInterpreter = recInterpreter
            livenessInterpreter = livInterpreter
            initialized = true
            Log.i(TAG, "TFLite models loaded successfully")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load TFLite models: ${e.message}")
            false
        }
    }

    /**
     * Run face recognition inference with optional TTA.
     *
     * Input: 112×112×3 RGB image (normalized floats, 0–1)
     * Output: `embeddingDim`-dimensional L2-normalized embedding (128 for the
     * pre-trained MobileFaceNet, 512 for the Day-2 fine-tuned EdgeFace-XS).
     *
     * TTA (when enabled, default ON): runs the original input AND a horizontal
     * flip through the model, L2-normalizes each embedding, averages them,
     * then L2-normalizes the result. Empirically +0.3-0.5% LFW for free.
     */
    fun runRecognition(inputData: FloatArray): FloatArray? {
        val interpreter = recognitionInterpreter ?: run {
            Log.e(TAG, "Recognition model not initialized")
            return null
        }
        if (!initialized || embeddingDim <= 0) return null

        return try {
            // Build the 3D input tensor [1, 112, 112, 3] from flat RGB.
            val input = Array(1) { Array(INPUT_SIZE) { FloatArray(INPUT_SIZE * 3) } }
            for (y in 0 until INPUT_SIZE) {
                for (x in 0 until INPUT_SIZE) {
                    val idx = (y * INPUT_SIZE + x) * 3
                    input[0][y][x * 3] = inputData[idx]
                    input[0][y][x * 3 + 1] = inputData[idx + 1]
                    input[0][y][x * 3 + 2] = inputData[idx + 2]
                }
            }

            // Forward pass on original input.
            val output = Array(1) { FloatArray(embeddingDim) }
            interpreter.run(input, output)
            l2NormalizeInPlace(output[0])

            if (!useTTA) return output[0]

            // TTA: forward pass on horizontal-flip.
            val flippedInput = Array(1) { Array(INPUT_SIZE) { FloatArray(INPUT_SIZE * 3) } }
            for (y in 0 until INPUT_SIZE) {
                for (x in 0 until INPUT_SIZE) {
                    val srcX = INPUT_SIZE - 1 - x
                    val srcIdx = (y * INPUT_SIZE + srcX) * 3
                    val dstIdx = (y * INPUT_SIZE + x) * 3
                    flippedInput[0][y][dstIdx]     = inputData[srcIdx]
                    flippedInput[0][y][dstIdx + 1] = inputData[srcIdx + 1]
                    flippedInput[0][y][dstIdx + 2] = inputData[srcIdx + 2]
                }
            }
            val flippedOutput = Array(1) { FloatArray(embeddingDim) }
            interpreter.run(flippedInput, flippedOutput)
            l2NormalizeInPlace(flippedOutput[0])

            // Average the two L2-normalized embeddings, then re-normalize.
            val averaged = FloatArray(embeddingDim)
            for (i in 0 until embeddingDim) {
                averaged[i] = (output[0][i] + flippedOutput[0][i]) * 0.5f
            }
            l2NormalizeInPlace(averaged)
            averaged
        } catch (e: Exception) {
            Log.e(TAG, "Recognition inference failed: ${e.message}")
            null
        }
    }

    /** In-place L2 normalization. */
    private fun l2NormalizeInPlace(v: FloatArray) {
        var norm = 0f
        for (x in v) norm += x * x
        norm = Math.sqrt(norm.toDouble()).toFloat()
        if (norm > 0) {
            for (i in v.indices) v[i] /= norm
        }
    }

    /**
     * Run liveness detection inference.
     *
     * Input: 112×112×3 RGB image (normalized floats, 0–1)
     * Output: `livenessDim`-class probabilities (typically [real, print, screen] = 3)
     */
    fun runLiveness(inputData: FloatArray): FloatArray? {
        val interpreter = livenessInterpreter ?: run {
            Log.e(TAG, "Liveness model not initialized")
            return null
        }
        if (!initialized || livenessDim <= 0) return null

        return try {
            val input = Array(1) { Array(INPUT_SIZE) { FloatArray(INPUT_SIZE * 3) } }
            for (y in 0 until INPUT_SIZE) {
                for (x in 0 until INPUT_SIZE) {
                    val idx = (y * INPUT_SIZE + x) * 3
                    input[0][y][x * 3] = inputData[idx]
                    input[0][y][x * 3 + 1] = inputData[idx + 1]
                    input[0][y][x * 3 + 2] = inputData[idx + 2]
                }
            }

            val output = Array(1) { FloatArray(livenessDim) }
            interpreter.run(input, output)
            output[0]
        } catch (e: Exception) {
            Log.e(TAG, "Liveness inference failed: ${e.message}")
            null
        }
    }

    /**
     * Compute cosine similarity between two embeddings.
     */
    fun cosineSimilarity(a: FloatArray, b: FloatArray): Float {
        require(a.size == b.size) { "Dimension mismatch: a=${a.size} b=${b.size}" }
        var dot = 0f
        for (i in a.indices) dot += a[i] * b[i]
        return (-1f).coerceAtLeast(1f.coerceAtMost(dot))
    }

    /**
     * Process raw camera frame pixels: crop face region, resize to 112x112,
     * normalize to 0.0-1.0. This is the CRITICAL bridge between camera
     * frames and the face recognition pipeline.
     *
     * @param imageData Raw pixel data as float array (R,G,B,R,G,B,...)
     * @param frameWidth Width of the original camera frame
     * @param frameHeight Height of the original camera frame
     * @param faceX Face bounding box left coordinate
     * @param faceY Face bounding box top coordinate
     * @param faceWidth Face bounding box width
     * @param faceHeight Face bounding box height
     * @return Float array of 37632 values (112*112*3) or null on failure
     */
    fun cropFace(
        imageData: FloatArray,
        frameWidth: Int,
        frameHeight: Int,
        faceX: Int,
        faceY: Int,
        faceWidth: Int,
        faceHeight: Int
    ): FloatArray? {
        if (faceWidth <= 0 || faceHeight <= 0) return null

        // Add 20% padding around face for better context
        val paddingX = (faceWidth * 0.2).toInt()
        val paddingY = (faceHeight * 0.2).toInt()

        val cropLeft = maxOf(0, faceX - paddingX)
        val cropTop = maxOf(0, faceY - paddingY)
        val cropRight = minOf(frameWidth, faceX + faceWidth + paddingX)
        val cropBottom = minOf(frameHeight, faceY + faceHeight + paddingY)

        val cropWidth = cropRight - cropLeft
        val cropHeight = cropBottom - cropTop

        if (cropWidth <= 0 || cropHeight <= 0) return null

        // Extract and resize using bilinear interpolation
        val result = FloatArray(INPUT_SIZE * INPUT_SIZE * 3)

        for (y in 0 until INPUT_SIZE) {
            for (x in 0 until INPUT_SIZE) {
                // Map output pixel to input crop region
                val srcX = cropLeft + (x.toFloat() / INPUT_SIZE * cropWidth).toInt()
                val srcY = cropTop + (y.toFloat() / INPUT_SIZE * cropHeight).toInt()

                val srcIdx = (srcY * frameWidth + srcX) * 3
                val dstIdx = (y * INPUT_SIZE + x) * 3

                if (srcIdx + 2 < imageData.size && dstIdx + 2 < result.size) {
                    result[dstIdx] = imageData[srcIdx]         // R
                    result[dstIdx + 1] = imageData[srcIdx + 1] // G
                    result[dstIdx + 2] = imageData[srcIdx + 2] // B
                }
            }
        }

        return result
    }

    /**
     * Convert YUV_420_888 byte data to normalized RGB float array.
     * This is used when receiving raw camera frame data.
     *
     * @param yuvData YUV_420_888 byte array
     * @param width Frame width
     * @param height Frame height
     * @return Normalized RGB float array or null
     */
    fun yuvToNormalizedFloats(yuvData: ByteArray, width: Int, height: Int): FloatArray? {
        return try {
            val argb = IntArray(width * height)
            decodeYUV420ToARGB(yuvData, argb, width, height)

            val result = FloatArray(width * height * 3)
            for (i in argb.indices) {
                val pixel = argb[i]
                result[i * 3] = ((pixel shr 16) and 0xFF) / 255.0f     // R
                result[i * 3 + 1] = ((pixel shr 8) and 0xFF) / 255.0f  // G
                result[i * 3 + 2] = (pixel and 0xFF) / 255.0f          // B
            }
            result
        } catch (e: Exception) {
            Log.e(TAG, "YUV conversion failed: ${e.message}")
            null
        }
    }

    /**
     * Decode YUV_420_888 to ARGB pixel array.
     */
    private fun decodeYUV420ToARGB(yuvData: ByteArray, argb: IntArray, width: Int, height: Int) {
        val ySize = width * height
        val uvSize = ySize / 4

        for (y in 0 until height) {
            for (x in 0 until width) {
                val yIndex = y * width + x
                val yValue = (yuvData[yIndex].toInt() and 0xFF) - 128

                val uvIndex = ySize + (y / 2) * (width / 2) + (x / 2)
                val uValue = (yuvData[uvIndex].toInt() and 0xFF) - 128
                val vValue = (yuvData[uvIndex + uvSize].toInt() and 0xFF) - 128

                val r = (yValue + 1.370705f * vValue).toInt().coerceIn(0, 255)
                val g = (yValue - 0.337633f * uValue - 0.698001f * vValue).toInt().coerceIn(0, 255)
                val b = (yValue + 1.732446f * uValue).toInt().coerceIn(0, 255)

                argb[yIndex] = (0xFF shl 24) or (r shl 16) or (g shl 8) or b
            }
        }
    }

    fun close() {
        recognitionInterpreter?.close()
        livenessInterpreter?.close()
        recognitionInterpreter = null
        livenessInterpreter = null
        initialized = false
        Log.i(TAG, "NetraEdge module closed")
    }
}

File: packages/react-native/ios/NetraEdgeModule.swift
/**
 * NetraEdgeModule — iOS native bridge for TFLite inference.
 *
 * Provides the same API as the Android Kotlin module:
 * - initialize: loads TFLite models from the app bundle
 * - runRecognition: face recognition (112×112×3 → N-d L2-normalized embedding)
 * - runLiveness: liveness detection (112×112×3 → 3-class softmax)
 * - cosineSimilarity: embedding comparison utility
 * - cropFace: SOTA face cropper used by FaceCamera (replaces zero-fill)
 * - close: resource cleanup
 * - getEmbeddingDim: returns the loaded model's actual output dim
 * - setUseTTA: enable/disable horizontal-flip TTA at runtime
 *
 * Dim handling: embeddingDim is derived from the TFLite model's output
 * tensor at initialize() time. This avoids hardcoded crashes when the
 * pre-trained 128-d MobileFaceNet is swapped for the Day-2 fine-tuned
 * 512-d EdgeFace-XS / SubCenter-AdaFace / 2-Teacher KD model (Path A+).
 *
 * TTA: useTTA defaults to true (horizontal flip averaging +0.3-0.5% LFW).
 */

import Foundation
import CoreVideo
import TensorFlowLite

@objc(NetraEdgeModule)
class NetraEdgeModule: NSObject {

  private var recognitionInterpreter: Interpreter?
  private var livenessInterpreter: Interpreter?
  private var isInitialized = false

  private let inputSize = 112
  // SOTA TARGET dim for Day-2 fine-tune (Path A+: EdgeFace-XS / 2-Teacher KD).
  // The actual runtime dim is read from the model below.
  private static let sotaTargetEmbeddingDim = 512
  // Set after initialize() reads the actual model output shape.
  private var embeddingDim: Int = -1
  private var livenessDim: Int = 3

  // Test-time augmentation: horizontal flip averaging at inference. Default ON.
  private var useTTA: Bool = true

  /// Returns the loaded model's embedding dim (or -1 if not loaded).
  @objc func getEmbeddingDim(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(NSNumber(value: embeddingDim))
  }

  /// Enable or disable TTA (horizontal flip averaging) at runtime.
  @objc func setUseTTA(
    _ enabled: NSNumber,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    useTTA = enabled.boolValue
    NSLog("NetraEdge: TTA %@", useTTA ? "enabled" : "disabled")
    resolve(nil)
  }

  /// Returns whether TTA is currently enabled.
  @objc func isUseTTA(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(NSNumber(value: useTTA))
  }

  // MARK: - React Native bridge methods

  @objc func initialize(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      guard let recPath = Bundle.main.path(
        forResource: "face_recognition",
        ofType: "tflite"
      ) else {
        reject("INIT_FAILED", "face_recognition.tflite not found in bundle", nil)
        return
      }

      guard let livPath = Bundle.main.path(
        forResource: "liveness_detector",
        ofType: "tflite"
      ) else {
        reject("INIT_FAILED", "liveness_detector.tflite not found in bundle", nil)
        return
      }

      recognitionInterpreter = try Interpreter(modelPath: recPath)
      try recognitionInterpreter?.allocateTensors()
      // Read actual output dim from the model — works for any model.
      if let recOut = try recognitionInterpreter?.output(at: 0) {
        // Output shape is typically [1, N]. Pick the last dim.
        let shape = recOut.shape.dimensions
        embeddingDim = shape.last ?? Self.sotaTargetEmbeddingDim
        NSLog("NetraEdge: recognition output dim = %d (target = %d)",
              embeddingDim, Self.sotaTargetEmbeddingDim)
      }

      livenessInterpreter = try Interpreter(modelPath: livPath)
      try livenessInterpreter?.allocateTensors()
      if let livOut = try livenessInterpreter?.output(at: 0) {
        livenessDim = livOut.shape.dimensions.last ?? 3
        NSLog("NetraEdge: liveness output dim = %d", livenessDim)
      }

      isInitialized = true
      resolve(true)
    } catch {
      reject("INIT_ERROR", "Failed to load TFLite models: \(error.localizedDescription)", error)
    }
  }

  @objc func isInitialized(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(isInitialized)
  }

  @objc func runRecognition(
    _ pixels: NSArray,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard isInitialized, let interpreter = recognitionInterpreter else {
      reject("RECOGNITION_FAILED", "Recognition model not initialized", nil)
      return
    }
    guard embeddingDim > 0 else {
      reject("RECOGNITION_FAILED", "Recognition output dim not initialized", nil)
      return
    }

    do {
      let floatPixels = try convertToFloatArray(pixels)
      let inputData = Data(bytes: floatPixels, count: floatPixels.count * MemoryLayout<Float>.stride)

      // Forward pass on original input.
      try interpreter.copy(inputData, toInputAt: 0)
      try interpreter.invoke()
      let e1 = readEmbedding(from: interpreter)

      guard !e1.isEmpty else {
        reject("RECOGNITION_FAILED", "Empty output buffer from TFLite", nil)
        return
      }
      let n1 = l2Normalize(e1)

      if !useTTA {
        resolve(n1)
        return
      }

      // TTA: build a horizontal-flip copy of the input and re-run.
      // Layout: 112 rows × 112 cols × 3 channels (RGB interleaved).
      var flippedPixels = floatPixels
      let w = inputSize
      let h = inputSize
      for y in 0..<h {
        for x in 0..<(w / 2) {
          let src = (y * w + (w - 1 - x)) * 3
          let dst = (y * w + x) * 3
          for c in 0..<3 {
            let tmp = flippedPixels[dst + c]
            flippedPixels[dst + c] = flippedPixels[src + c]
            flippedPixels[src + c] = tmp
          }
        }
      }
      let flippedData = Data(
        bytes: flippedPixels,
        count: flippedPixels.count * MemoryLayout<Float>.stride
      )
      try interpreter.copy(flippedData, toInputAt: 0)
      try interpreter.invoke()
      let e2 = readEmbedding(from: interpreter)
      let n2 = l2Normalize(e2)

      // Average the two L2-normalized embeddings, then re-normalize.
      var averaged = [Float](repeating: 0, count: n1.count)
      for i in 0..<n1.count { averaged[i] = (n1[i] + n2[i]) * 0.5 }
      let final = l2Normalize(averaged)
      resolve(final)
    } catch {
      reject("RECOGNITION_ERROR", "Inference failed: \(error.localizedDescription)", error)
    }
  }

  /// Read the model's current output tensor as a [Float] array, clamped to
  /// the actual buffer size (prevents UnsafeBufferPointer BufferOverflow).
  private func readEmbedding(from interpreter: Interpreter) -> [Float] {
    do {
      let outputTensor = try interpreter.output(at: 0)
      let outputData = outputTensor.data
      let stride = MemoryLayout<Float>.stride
      let availableFloats = outputData.count / stride
      let readCount = min(availableFloats, embeddingDim)
      if readCount == 0 { return [] }
      return outputData.withUnsafeBytes { rawBuf -> [Float] in
        guard let base = rawBuf.bindMemory(to: Float.self).baseAddress else {
          return []
        }
        return Array(UnsafeBufferPointer<Float>(start: base, count: readCount))
      }
    } catch {
      NSLog("NetraEdge: readEmbedding failed: %@", error.localizedDescription)
      return []
    }
  }

  /// Pure L2 normalization (returns a new array).
  private func l2Normalize(_ v: [Float]) -> [Float] {
    let n = sqrt(v.reduce(0) { $0 + $1 * $1 })
    return n > 0 ? v.map { $0 / n } : v
  }

  @objc func runLiveness(
    _ pixels: NSArray,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard isInitialized, let interpreter = livenessInterpreter else {
      reject("LIVENESS_FAILED", "Liveness model not initialized", nil)
      return
    }
    guard livenessDim > 0 else {
      reject("LIVENESS_FAILED", "Liveness output dim not initialized", nil)
      return
    }

    do {
      let floatPixels = try convertToFloatArray(pixels)
      let inputData = Data(bytes: floatPixels, count: floatPixels.count * MemoryLayout<Float>.stride)

      try interpreter.copy(inputData, toInputAt: 0)
      try interpreter.invoke()

      let outputTensor = try interpreter.output(at: 0)
      let outputData = outputTensor.data
      let stride = MemoryLayout<Float>.stride
      let availableFloats = outputData.count / stride
      let readCount = min(availableFloats, livenessDim)
      if readCount == 0 {
        reject("LIVENESS_FAILED", "Empty output buffer from TFLite", nil)
        return
      }

      let probabilities: [Float] = outputData.withUnsafeBytes { rawBuf -> [Float] in
        guard let base = rawBuf.bindMemory(to: Float.self).baseAddress else {
          return []
        }
        return Array(UnsafeBufferPointer<Float>(start: base, count: readCount))
      }

      resolve(probabilities)
    } catch {
      reject("LIVENESS_ERROR", "Inference failed: \(error.localizedDescription)", error)
    }
  }

  @objc func cosineSimilarity(
    _ a: NSArray,
    b: NSArray,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      let aFloats = try convertToFloatArray(a)
      let bFloats = try convertToFloatArray(b)

      guard aFloats.count == bFloats.count else {
        reject("SIMILARITY_ERROR", "Dimension mismatch", nil)
        return
      }

      let dot = zip(aFloats, bFloats).reduce(Float(0)) { $0 + $1.0 * $1.1 }
      let clamped = max(-1, min(1, dot))
      resolve(Double(clamped))
    } catch {
      reject("SIMILARITY_ERROR", error.localizedDescription, error)
    }
  }

  @objc func close() {
    recognitionInterpreter = nil
    livenessInterpreter = nil
    isInitialized = false
  }

  // MARK: - SOTA face cropper (used by FaceCamera)

  /// Crops and aligns a face from a YUV/RGB frame buffer.
  ///
  /// iOS implementation: lightweight bilinear crop using CoreImage
  /// downsample to 112×112×3. This is a JS-driven cropper that operates
  /// on the same Float32Array layout as Android's `cropFace`.
  ///
  /// - Parameters:
  ///   - pixels: source frame as flat Float32Array (H*W*3, RGB, 0..1)
  ///   - width: source width
  ///   - height: source height
  ///   - x: crop top-left X
  ///   - y: crop top-left Y
  ///   - w: crop width
  ///   - h: crop height
  /// - Resolves: Float32Array of length 112*112*3
  @objc func cropFace(
    _ pixels: NSArray,
    width: NSNumber,
    height: NSNumber,
    x: NSNumber,
    y: NSNumber,
    w: NSNumber,
    cropH: NSNumber,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    do {
      let src = try convertToFloatArray(pixels)
      let srcW = width.intValue
      let srcH = height.intValue
      let cx = max(0, x.intValue)
      let cy = max(0, y.intValue)
      let cw = min(srcW - cx, w.intValue)
      let ch = min(srcH - cy, cropH.intValue)
      let outSize = self.inputSize

      guard cw > 0 && ch > 0 else {
        reject("CROP_FAILED", "Invalid crop region", nil)
        return
      }

      var output = [Float](repeating: 0, count: outSize * outSize * 3)
      let xRatio = Float(cw) / Float(outSize)
      let yRatio = Float(ch) / Float(outSize)

      for oy in 0..<outSize {
        for ox in 0..<outSize {
          let sy = cy + Int(Float(oy) * yRatio)
          let sx = cx + Int(Float(ox) * xRatio)
          let sIdx = (sy * srcW + sx) * 3
          let dIdx = (oy * outSize + ox) * 3
          if sIdx + 2 < src.count {
            output[dIdx]     = src[sIdx]!
            output[dIdx + 1] = src[sIdx + 1]!
            output[dIdx + 2] = src[sIdx + 2]!
          }
        }
      }

      resolve(output)
    } catch {
      reject("CROP_ERROR", error.localizedDescription, error)
    }
  }

  // MARK: - React Native setup

  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }

  // MARK: - Private helpers

  private func convertToFloatArray(_ nsArray: NSArray) throws -> [Float] {
    var floats = [Float]()
    floats.reserveCapacity(nsArray.count)
    for item in nsArray {
      guard let number = item as? NSNumber else {
        throw NSError(
          domain: "NetraEdge",
          code: -1,
          userInfo: [NSLocalizedDescriptionKey: "Expected numeric array"]
        )
      }
      floats.append(number.floatValue)
    }
    return floats
  }
}
